# HG changeset patch
# User alta88@fixall.com
# Date 1745089972 -7200
# Parent  0bca544d0a32dd37e01fe2dbc0d1b0cdd3d1be56
NNN24 - No unnecessary tooltips in threadpane columns, restore nsITree behavior.
- For both table view and card view.

Authored by Alta88 and contributed to the Betterbird Project.
See https://github.com/Betterbird/thunderbird-patches/blob/main/LICENSE for license details.
All uses require attribution to the Author.

Thunderbird bug https://bugzilla.mozilla.org/show_bug.cgi?id=1914128

diff --git a/mail/base/content/about3Pane.js b/mail/base/content/about3Pane.js
--- a/mail/base/content/about3Pane.js
+++ b/mail/base/content/about3Pane.js
@@ -4993,8 +4993,9 @@ var threadPane = {
     threadTree.addEventListener("expanded", this);
     threadTree.addEventListener("collapsed", this);
     threadTree.addEventListener("scroll", this);
     threadTree.addEventListener("showplaceholder", this);
+    threadTree.addEventListener("mouseover", this);
   },
 
   uninit() {
     Services.prefs.removeObserver("mailnews.tags.", this);
@@ -5044,8 +5045,11 @@ var threadPane = {
         break;
       case "drop":
         this._onDrop(event);
         break;
+      case "mouseover":
+        this._onMouseOver(event);
+        break;
       case "expanded":
       case "collapsed":
         if (event.detail == threadTree.selectedIndex) {
           // The selected index hasn't changed, but a collapsed row represents
@@ -5847,8 +5851,40 @@ var threadPane = {
 
     messageEditor.updateThreading(action, newParentMessageURI);
   },
 
+  _onMouseOver(event) {
+    // Prevent tooltips if unnecessary, ie text is not overflowing/in ellipsis.
+    const titleNode = event.target?.title
+      ? event.target
+      : event.target?.closest("td");
+    if (!titleNode?.title) {
+      return;
+    }
+
+    const textNode = titleNode.querySelector("span") ?? titleNode;
+    if (titleNode.title != textNode.textContent) {
+      return;
+    }
+
+    // Subject in table view is special.
+    const subjectNode = titleNode.querySelector(".subject-line");
+    const textElementNode = subjectNode ?? textNode;
+    const textContainerNode =
+      subjectNode ??
+      (titleNode.clientWidth ? titleNode : titleNode.parentElement);
+    if (textElementNode.scrollWidth > textContainerNode.clientWidth) {
+      return;
+    }
+
+    titleNode.title = "";
+    titleNode.addEventListener(
+      "mouseout",
+      () => (titleNode.title = textNode.textContent),
+      { once: true }
+    );
+  },
+
   _jsTree: {
     QueryInterface: ChromeUtils.generateQI(["nsIMsgJSTree"]),
     _inBatch: false,
     beginUpdateBatch() {
diff --git a/mail/themes/shared/mail/tree-listbox.css b/mail/themes/shared/mail/tree-listbox.css
--- a/mail/themes/shared/mail/tree-listbox.css
+++ b/mail/themes/shared/mail/tree-listbox.css
@@ -657,12 +657,13 @@ table[is="tree-view-table"] {
 
   td {
     max-width: 0;
     height: inherit;
-    overflow-x: clip;
+    overflow-x: hidden;
     text-overflow: ellipsis;
     white-space: nowrap;
-    padding: 0 6px;
+    padding: 0;
+    padding-inline: 6px 0;
     position: relative;
     user-select: none;
     box-sizing: border-box;
 
